wip: fix(plugin): enforce strict versioning for static plugins

gnosek opened the pull request:
Statically linked plugins have stricter version requirements, since they rely on the layout of the `plugin_api` struct. So, when loading a statically linked plugin, we enforce the plugin API to be exactly equal (up to the patch level) to the framework API.

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines in the [CONTRIBUTING.md](https://github.com/falcosecurity/.github/blob/master/CONTRIBUTING.md) file and learn how to compile Falco from source [here](https://falco.org/docs/source).
2. Please label this pull request according to what type of issue you are addressing.
3. Please add a release note!
4. If the PR is unfinished while opening it specify a wip in the title before the actual title, for example, "wip: my awesome feature"
-->

**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

/kind bug

> /kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

> /kind feature

<!--
Please remove the leading whitespace before the `/kind <>` you uncommented.
-->

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area API-version

> /area build

> /area CI

> /area driver-kmod

> /area driver-bpf

> /area driver-modern-bpf

> /area libscap-engine-bpf

> /area libscap-engine-gvisor

> /area libscap-engine-kmod

> /area libscap-engine-modern-bpf

> /area libscap-engine-nodriver

> /area libscap-engine-noop

> /area libscap-engine-source-plugin

> /area libscap-engine-savefile

> /area libscap

> /area libpman

/area libsinsp

> /area tests

> /area proposals

<!--
Please remove the leading whitespace before the `/area <>` you uncommented.
-->

**Does this PR require a change in the driver versions?**

> /version driver-API-version-major

> /version driver-API-version-minor

> /version driver-API-version-patch

> /version driver-SCHEMA-version-major

> /version driver-SCHEMA-version-minor

> /version driver-SCHEMA-version-patch

<!--
Please remove the leading whitespace before the `/version <>` you uncommented.
-->

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

<!--
Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
If PR is `kind/failing-tests` please post the related issues/tests in a comment and do not use `Fixes`.
-->

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

<!--
If no, you have to do nothing.
If yes, a release note is required:
Delete `NONE` and enter your extended release note in the block below.
Please note, the release note follows the "conventional commit specification" (https://www.conventionalcommits.org/en/v1.0.0/):
For example: `fix: broken link`.
If the PR requires additional action from users switching to the new release, prepend the string "action required:".
For example, `action required: change the API interface of libscap`.
-->

```release-note
NONE
```

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/2278#" title="Author self-approved">gnosek</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [gnosek]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

leogr commented:
/milestone 0.21.0

gnosek committed:
new(sinsp): add tools to manage plugin ABI compatibility

`extract_plugin_abi.py` parses the output from pahole describing
the `plugin_api` struct (with the offsets and sizes of all fields)
and generates a `plugin_abi.h` header with static asserts to ensure
the ABI is up to date (on x86_64/Linux).

`diff_plugin_abi.py` takes two copies of the generated plugin_abi.h
files (presumably, one from the repo and a freshly generated one)
and raises an error if:
 * a struct member has been moved/removed without ABI version bump
 * the ABI version has been increased without breaking changes

Note: adding a field at the end of the plugin_api struct does not
count as a breaking change.

Signed-off-by: Grzegorz Nosek <grzegorz.nosek@sysdig.com>

gnosek committed:
chore(sinsp): generate and include plugin_abi.h

The current ABI is arbitrarily chosen as v3 (to match the major API
version).

Note: when making incompatible ABI changes, this file will prevent
the build from succeeding. When you're making ABI changes on purpose,
the easiest way is to remove everything but the ABI version `#define`
from plugin_abi.h and once everything builds, regenerate plugin_abi.h

(see tools/plugin_abi/extract_plugin_abi.py)

Signed-off-by: Grzegorz Nosek <grzegorz.nosek@sysdig.com>

gnosek committed:
new(sinsp): support explicit ABI versioning

Introduce an explicit ABI version field to plugin_api, as well as
a size field. This allows us to explicitly detect incompatible
ABIs, but also to detect struct changes within an ABI version.

The size field is actually pretty important (maybe more than
the actual version field, as long as get_api_version can be easily
found), because it's the only way to know how big the plugin-provided
struct actually is. With plugins targeting older API versions,
we're doing a basically guaranteed out of bounds read otherwise
(if the plugin_api struct in the framework is larger than what
the plugin exports).

Signed-off-by: Grzegorz Nosek <grzegorz.nosek@sysdig.com>

gnosek committed:
fix(build): don't require plugin_abi.h to be formatted

Signed-off-by: Grzegorz Nosek <grzegorz.nosek@sysdig.com>

FedeDP commented:
Moving to next milestone, no need to hurry for the current one.
/milestone 0.22.0

leogr commented:
Hey @gnosek 

any update on this?
